src/backend/distributed/planner/query_colocation_checker.c

```diff
@@ -0,0 +1,303 @@
+/*-------------------------------------------------------------------------
+ *
+ * query_colocation_checker.c implements the logic for determining
+ * whether any subqueries in a given query are co-located (e.g.,
+ * distribution keys of the relations inside subqueries are equal).
+ *
+ * Copyright (c) 2018, Citus Data, Inc.
```

Comment:
It would be nice to add a paragraph about the details of the implementation here (mainly, checking against anchor RTE, wrapping tables in a subquery).